wgsl/index.bs

```diff
@@ -5480,6 +5479,74 @@ See [[#sync-builtin-functions]].
     <td>Bitwise-exclusive-or. [=Component-wise=] when |T| is a vector.
 </table>
 
+
+<table class='data'>
+  <caption>Bit shift expressions</caption>
+  <thead>
+    <tr><th>Precondition<th>Conclusion<th>Notes
+  </thead>
+
+  <tr algorithm="concrete shift left">
+    <td>|e1|: |T|<br>
+    |e2|: |TS|<br>
+    |S| is [INT]<br>
+    |T| is |S| or vec|N|&lt;|S|&gt;<br>
+    |TS| is u32 when |T| is |S|, otherwise |TS| is vec|N|&lt;u32&gt;
+    <td class="nowrap">|e1| `<<` |e2|: |T|
+    <td>Shift left (concrete):
+
+        Shift |e1| left, inserting zero bits at the least significant positions,
+        and discarding the most significant bits.
+
+        The number of bits to shift is the value of |e2|, modulo the bit width of |e1|.<br>
+
+        [=Component-wise=] when |T| is a vector.
+
+  <tr algorithm="abstract shift left">
+    <td>|e1|: |T|<br>
+    |e2|: |T|<br>
+    |T| is AbstractInt or vec|N|&lt;AbstractInt&gt;
+    <td class="nowrap">|e1| `<<` |e2|: |T|
+    <td>Shift left (abstract):
+
+        Shift |e1| left, inserting zero bits at the least significant positions,
+        and discarding the most significant bits.
+
+        The number of bits to shift is the value of |e2|.
+
+        [=Component-wise=] when |T| is a vector.
+
+        It is a [=shader-creation error=] if the calculation would overflow the AbstractInt type.
+        Overflow occurs when both 0 and 1 bits would be discarded,
+        or when the most significant bit of |e1| would differ from the most
+        significant bit of the result.
+
+        It is a [=shader-creation error=] if any of the |e2| values are less than 0.
+
+  <tr algorithm="shift right">
+    <td>|e1|: |T|<br>
+    |e2|: |TS|<br>
+    [ALLINTEGRALDECL]<br>
+    |TS| is vec|N|&lt;AbstractInt&gt; or vec|N|&lt;u32&gt; when |T| is vec|N|&lt;S&gt;,
+    otherwise |TS| is AbstractInt or u32
+    <td class="nowrap">|e1| >> |e2|: |T|
+    <td>Shift right.
+
+        If |e1| is unsigned [=concrete=], shift |e1| right, inserting zero bits at the most significant positions,
+        and discarding the least significant bits.
+
+        If |e1| is AbstractInt or signed [=concrete=], shift |e1| right, copying the sign bit of |e1| into the most significant positions,
+        and discarding the least significant bits.
+
+        The number of bits to shift is the value of |e2|.
+        If |e1| has a [=concrete=] type, the shift value is modulo the bit width of |e1|.<br>
```

Comment:
No. We don't want to rely on AbstractInt having a specific bit width.

If you do  1 >> 65 then you get zero.
If you do 1 << 65 you get shader creation error.
If someone makes an extension where AbstractInt has 128 bits, then 1 << 64 can become meaningful, and strictly more programs become valid. But old programs that are valid still have the same meaning.

Reply:
Right and the need to specify this for concrete numbers is because they _could_ happen in the shader where it may modulo. But an abstract number happens compile time and we can make it 0.